ssl_it_test.go

```diff
@@ -0,0 +1,246 @@
+	MutualAuthenticationStatusRequired MutualAuthenticationStatus = "REQUIRED"
+)
+
+func mutualAuthenticationTests(t *testing.T) {
+	testCases := []struct {
+		name          string
+		muStatus      MutualAuthenticationStatus
+		ca, cert, key string
+		wantErr       bool
+	}{
+		{
+			// ported from, com.hazelcast.client.nio.ssl.ClientAuthenticationTest.whenRequired
+			name:     "whenRequired",
+			muStatus: MutualAuthenticationStatusRequired,
+			ca:       "testdata/openssl/server/server.crt",
+			cert:     "testdata/openssl/client/client.crt",
+			key:      "testdata/openssl/client/client.key",
+			wantErr:  false,
+		},
+		{
+			// ported from, com.hazelcast.client.nio.ssl.ClientAuthenticationTest#whenRequired_andServerNotAuthenticated
+			name:     "whenRequired_andServerNotAuthenticated",
+			muStatus: MutualAuthenticationStatusRequired,
+			ca:       "testdata/unknown-server-cert.pem",
+			cert:     "testdata/openssl/client/client.crt",
+			key:      "testdata/openssl/client/client.key",
+			wantErr:  true,
+		},
+		{
+			// ported from, com.hazelcast.client.nio.ssl.ClientAuthenticationTest#whenRequired_andWrongClientAuthenticated
+			name:     "whenRequired_andWrongClientAuthenticated",
+			muStatus: MutualAuthenticationStatusRequired,
+			ca:       "testdata/openssl/server/server.crt",
+			cert:     "testdata/unknown-client-cert.pem",
+			key:      "testdata/unknown-client-key.pem",
+			wantErr:  true,
+		},
+		{
+			// both server and client are not known by the root certificate authority
+			name:     "whenRequired_BothClientAndServerUnknownByRootCA",
+			muStatus: MutualAuthenticationStatusRequired,
+			ca:       "testdata/unknown-server-cert.pem",
+			cert:     "testdata/unknown-client-cert.pem",
+			key:      "testdata/unknown-client-key.pem",
+			wantErr:  true,
+		},
+		{
+			// ported from, com.hazelcast.client.nio.ssl.ClientAuthenticationTest#whenOptional
+			name:     "whenOptional",
+			muStatus: MutualAuthenticationStatusOptional,
+			ca:       "testdata/openssl/server/server.crt",
+			cert:     "testdata/openssl/client/client.crt",
+			key:      "testdata/openssl/client/client.key",
+			wantErr:  false,
+		},
+		{
+			// ported from, com.hazelcast.client.nio.ssl.ClientAuthenticationTest#whenOptional_andServerNotAuthenticated
+			name:     "whenOptional_andServerNotAuthenticated",
+			muStatus: MutualAuthenticationStatusOptional,
+			ca:       "testdata/unknown-server-cert.pem",
+			cert:     "testdata/openssl/client/client.crt",
+			key:      "testdata/openssl/client/client.key",
+			wantErr:  true,
+		},
+		//{
+		//	// ported from, com.hazelcast.client.nio.ssl.ClientAuthenticationTest#whenOptional_andWrongClientAuthenticated
+		//	name:     "whenOptional_andWrongClientAuthenticated",
+		//	muStatus: MutualAuthenticationStatusOptional,
+		//	ca:       "testdata/openssl/server/server.crt",
+		//	cert:     "testdata/unknown-client-cert.pem",
+		//	key:      "testdata/unknown-client-key.pem",
+		//	wantErr:  true,
+		//},
+		{
```

Comment:
Why was this test commented out?

Reply:
As I mentioned in PR comment;

`The test whenOptional_andWrongClientAuthenticated supposed to fail but it passes. For now, I commented down that test case in codebase, waiting for the to be reviewed.`

Reply:
Could you uncomment it?

Reply:
Uncommented, [e48b360](https://github.com/hazelcast/hazelcast-go-client/pull/859/commits/e48b360e7d97f6fcca8d8d8e204a4524c842e17b).